### PR TITLE
Implement #ab_seen and #ab_assigned in MockAdapter

### PR DESCRIPTION
**Pull request by urbanautomaton**
Hi there,

While switching to the mock adapter for test usage, we noticed that the `#ab_seen` and `#ab_assigned` methods from the abstract adapter weren't implemented in the mock adapter.

This PR introduces test coverage for the mock adapter, and adds implementations for the missing methods. I'd really appreciate a look-over to make sure I've got the meaning of these methods correct.

At the moment I've only added tests for the metric- and experiment-related methods, not the connection management ones. This because I think the former can probably be extracted and used to test all the adapters, whereas tests for the connection management behaviour are more likely to be adapter-specific. If you're happy to merge this, I'll have a stab at sharing the tests with the redis, mongo etc. adapters.

If there's anything you'd prefer done differently, or something I've missed, just let me know. :)

Cheers,
Simon

### CODE REVIEW

**Comment by phillbaker** [comment]
Thanks! Give me a day or two to look this over, but at first glance, this looks great. 

I've been thinking about shared tests for the adapters a bunch, which may help if someone wants to implement a new one. I was thinking of something similar to https://github.com/jnunemaker/flipper/blob/master/lib/flipper/spec/shared_adapter_specs.rb.

**Comment by phillbaker** [comment]
This looks great! Had a super minor nitpick, but I'll go ahead and merge this evening if you're busy.

**Comment by urbanautomaton** [comment]
Haha, nitpick away. 😄

File: lib/vanity/adapters/mock_adapter.rb [hunk 1]
```diff
@@ -134,17 +132,27 @@ def ab_not_showing(experiment, identity)
...
-        @experiments[experiment][:alternatives] ||= {}
-        alt = @experiments[experiment][:alternatives][alternative] ||= {}
+        alt = alternative(experiment, alternative)
         alt[:participants] ||= Set.new
         alt[:participants] << identity
       end
 
+      def ab_seen(experiment, identity, alternative)
+        if ab_assigned(experiment, identity) == alternative
+          return alternative
```
Comment (urbanautomaton) [inline lib/vanity/adapters/mock_adapter.rb]: D'oh - will correct, thanks.

Actually, I've been working on sharing the tests with the other adapters and I think I might have the method signature of this one wrong. Looking at the redis adapter it seems that for most methods having `alternative` as a parameter, they expect the alternative id. But for `ab_seen` it seems like it expects an instance of `Alternative`.

Compare `#ab_add_participant` to `#ab_seen` in the `RedisAdapter` [here](https://github.com/Futurelearn/vanity/blob/f221e16ebad0d32f74fa5c2a9ca764bcb9b1de66/lib/vanity/adapters/redis_adapter.rb#L172-L186):

```ruby
def ab_add_participant(experiment, alternative, identity)
  call_redis_with_failover(experiment, alternative, identity) do
    # interpolates `alternative` directly
    @experiments.sadd "#{experiment}:alts:#{alternative}:participants", identity
  end
end

def ab_seen(experiment, identity, alternative)
  call_redis_with_failover(experiment, identity, alternative) do
    # calls `#id` before interpolating
    if @experiments.sismember "#{experiment}:alts:#{alternative.id}:participants", identity
      alternative
    else
      nil
    end
  end
end
```

Is that right? If so I'll update the test and implementation...

Comment (phillbaker) [inline lib/vanity/adapters/mock_adapter.rb]: Hm, that looks right. ActiveRecordAdapter looks similar: https://github.com/Futurelearn/vanity/blob/f221e16ebad0d32f74fa5c2a9ca764bcb9b1de66/lib/vanity/adapters/active_record_adapter.rb#L260.

That's confusing 😒

Comment (urbanautomaton) [inline lib/vanity/adapters/mock_adapter.rb]: Yeah, it is a little. If the tests reflect this then it should at least be a bit better documented, I guess. I see from the docs that the adapter API is intended as a public extension point, so I guess we'd have to be pretty careful about "correcting" the method signature now it's out there.

Comment (urbanautomaton) [inline lib/vanity/adapters/mock_adapter.rb]: Well, I've updated the implementation and tests to match the other adapters (and remove the explicit return) - if we think the mismatch is too confusing we can address it in another PR perhaps?

Comment (phillbaker) [inline lib/vanity/adapters/mock_adapter.rb]: Yup, thanks!

**Pull request merged**